UnitsNet.Tests/QuantityComparisonTests.cs

```diff
@@ -0,0 +1,114 @@
+        {
+            var firstMass = Mass.FromGrams(0.001);
+            var secondMass = firstMass.ToUnit(MassUnit.Microgram);
+
+            Assert.Equal(firstMass, secondMass);
+            Assert.Equal(secondMass, firstMass);
+        }
+
+        [Fact]
+        public void Equals_WithVeryCloseQuantityInSameUnit_ReturnsFalse()
+        {
+            var firstMass = Mass.FromGrams(0);
+            var secondMass = Mass.FromGrams(double.Epsilon);
+
+            Assert.NotEqual(firstMass, secondMass);
+            Assert.NotEqual(secondMass, firstMass);
+        }
+
+        [Fact]
+        public void GetHasCode_WithSameQuantityInAnotherUnit_ReturnsSameValue()
+        {
+            var firstMass = Mass.FromGrams(0.001);
+            var secondMass = firstMass.ToUnit(MassUnit.Microgram);
+
+            Assert.Equal(firstMass.GetHashCode(), secondMass.GetHashCode());
+        }
+
+        [Fact]
+        public void GetHasCode_WithVeryCloseQuantities_ReturnsSameValue()
+        {
+            var firstMass = Mass.FromGrams(0);
+            var secondMass = Mass.FromGrams(double.Epsilon);
+
+            Assert.Equal(firstMass.GetHashCode(), secondMass.GetHashCode());
+        }
+
+        [Fact]
+        public void Contains_CollectionWithMixedUnits_ReturnsTrueIfValuesAreStrictlyEqual()
+        {
+            var firstMass = Mass.FromGrams(0.001);
+            var secondMass = firstMass.ToUnit(MassUnit.Microgram);
+
+            var collection = new HashSet<Mass> {firstMass};
+
+            Assert.Contains(firstMass, collection);
+            Assert.Contains(secondMass, collection);
+        }
+
+        [Fact]
+        public void Contains_CollectionWithVeryCloseQuantities_ReturnsTrueIfValuesAreStrictlyEqual()
+        {
+            var firstMass = Mass.FromGrams(0);
+            var secondMass = Mass.FromGrams(double.Epsilon);
+
+            var collection = new HashSet<Mass> {firstMass};
+
+            // equal hash codes do not imply object equality
+            Assert.Equal(firstMass.GetHashCode(), secondMass.GetHashCode());
+            Assert.NotEqual(firstMass, secondMass);
+
+            // Contains first checks HashCode, followed by Equals check
+            Assert.Contains(firstMass, collection);
+            Assert.DoesNotContain(secondMass, collection);
+        }
+
+        [Fact]
+        public void Sort_CollectionWithMixedUnits_SortedByNaturalOrder()
+        {
+            var firstMass = Mass.FromGrams(0);
+            var secondMass = Mass.FromGrams(double.Epsilon);
+            var thirdMass = Mass.FromGrams(0.001);
+            var fourthMass = thirdMass.ToUnit(MassUnit.Microgram);
+
+            var collection = new[] {thirdMass, firstMass, fourthMass, secondMass};
+            Array.Sort(collection);
+
+            Assert.Equal(firstMass, collection[0]);
+            Assert.Equal(secondMass, collection[1]);
+            Assert.Equal(thirdMass, collection[2]);
+            Assert.Equal(fourthMass, collection[3]);
```

Comment:
Is it co-incidental that third and fourth place are ordered this way, since they are equal?

Reply:
Oh, right, we are simply comparing equality so we don't actually know which value got inserted into 3rd or 4th position. We can't use `ReferenceEquals` either since these are struct values.

Reply:
I'm pretty sure it's in the order of insertion- but this might depend on the sorting algorithm.